README.md

# refextract

## About

A library for extracting references used in scholarly communication.

## Getting Started

Note: due to the usage of `mmap` resize functionality this library cannot be locally installed on a mac

### Docker Setup:

Before the first usage, or anytime a new library/dependency is changed a new docker image must be created using:
```shell
docker build --target refextract-tests -t refextract .
```

After that, spin up a `refextract` service with:
```shell
docker run -it -d -p 5000:5000 -v ./tests:/refextract/tests -v ./refextract:/refextract/refextract --name refextract refextract
```

### Running tests

Exec into the container via
```shell
docker exec -it refextract /bin/bash
```
Then simply run
```shell
pytest .
```

## Usage

To get structured information from a publication reference:


``` python
>>> from refextract import extract_journal_reference
>>> reference = extract_journal_reference('J.Phys.,A39,13445')
>>> print(reference)
{
'extra_ibids': [],
'is_ibid': False,
'misc_txt': '',
'page': '13445',
'title': 'J. Phys.',
'type': 'JOURNAL',
'volume': 'A39',
'year': '',

}
```

To extract references from a PDF:
``` python
>>> from refextract import extract_references_from_file
>>> references = extract_references_from_file('1503.07589.pdf')
>>> print(references[0])
{
'author': ['F. Englert and R. Brout'],
'doi': ['doi:10.1103/PhysRevLett.13.321'],
'journal_page': ['321'],
'journal_reference': ['Phys. Rev. Lett. 13 (1964) 321'],
'journal_title': ['Phys. Rev. Lett.'],
'journal_volume': ['13'],
'journal_year': ['1964'],
'linemarker': ['1'],
'raw_ref': ['[1] F. Englert and R. Brout, \u201cBroken symmetry and the mass of gauge vector mesons\u201d, Phys. Rev. Lett. 13 (1964) 321, doi:10.1103/PhysRevLett.13.321.'],
'texkey': ['Englert:1964et'],
'year': ['1964'],
}
```

To extract directly from a URL:
``` python
>>> from refextract import extract_references_from_url
>>> references = extract_references_from_url('https://arxiv.org/pdf/1503.07589.pdf')
>>> print(references[0])
{
'author': ['F. Englert and R. Brout'],
'doi': ['doi:10.1103/PhysRevLett.13.321'],
'journal_page': ['321'],
'journal_reference': ['Phys. Rev. Lett. 13 (1964) 321'],
'journal_title': ['Phys. Rev. Lett.'],
'journal_volume': ['13'],
'journal_year': ['1964'],
'linemarker': ['1'],
'raw_ref': ['[1] F. Englert and R. Brout, \u201cBroken symmetry and the mass of gauge vector mesons\u201d, Phys. Rev. Lett. 13 (1964) 321, doi:10.1103/PhysRevLett.13.321.'],
'texkey': ['Englert:1964et'],
'year': ['1964'],

}

```

## Notes
`refextract` depends on

[pdftotext](http://linux.die.net/man/1/pdftotext).

## Acknowledgments

`refextract` is based on code and ideas from the following people, who

contributed to the `docextract` module in Invenio:
- Alessio Deiana
- Federico Poli
- Gerrit Rindermann
- Graham R. Armstrong
- Grzegorz Szpura
- Jan Aage Lavik
- Javier Martin Montull
- Micha Moskovic
- Samuele Kaplun
- Thorsten Schwander
- Tibor Simko

## License
GPLv2


CHANGES.rst
..
   This file is part of refextract
   Copyright (C) 2015, 2016, 2017, 2018 CERN.

   refextract is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   refextract is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with refextract; if not, write to the Free Software Foundation, Inc.,
   59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

   In applying this license, CERN does not waive the privileges and immunities
   granted to it by virtue of its status as an Intergovernmental Organization
   or submit itself to any jurisdiction.


Changes
=======

Version 0.2.5 (2018-03-13)

- Handle all exceptions when extracting TeXkeys.

Version 0.2.4 (2018-02-28)

- Remove GarbageFullTextError.

Version 0.2.3 (2017-12-19)

- Handle all possible errors thrown by pyPDF2

- Fix normalization of CLIC report numbers.

Version 0.2.2 (2017-07-17)

- Handle pyPDF2 internal errors.

Version 0.2.1 (2017-07-02)

- Named destinations may not always have left and top coordinates. This case is
  now handled gracefully: no TeXkeys are returned by ``extract_texkeys_from_pdf``
  instead of raising an uncaught exception.

- Makes ``CFG_PATH_GFILE`` and ``CFG_PATH_PDFTOTEXT`` configurable through
  shell variables, with fallback on the output of ``which``, in order to allow
  for easier containerization.

Version 0.2.0 (2017-06-26)

- Substantial rewrite of the API. In particular:

  * ``extract_references_from_file``, ``extract_references_from_string``, and
    ``extract_references_from_url`` now return a list of the references,
    instead of an object with keys ``stats`` and ``references``.

  * If the number of TeXkeys that were extracted from the PDF metadata matches
    the number of references parsed by RefExtract, an extra ``texkey`` field is
    added to each returned reference.

  * The API now raises exceptions when it encounters an unrecoverable error.

  * Finally, the API now returns the list of raw references on which
    ``refextract`` worked.

Version 0.1.0 (2016-01-12)

- Initial export from Invenio Software <https://github.com/inveniosoftware/invenio>
- Restructured into stripped down, standalone version


refextract/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of refextract.
# Copyright (C) 2015, 2016 CERN.
#
# refextract is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# refextract is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with refextract; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization
# or submit itself to any jurisdiction.

"""Refextract."""

from __future__ import absolute_import, print_function, unicode_literals

from .references.api import (
    extract_journal_reference,
    extract_references_from_file,
    extract_references_from_string,
    extract_references_from_url,
)
from .version import __version__

__all__ = (
    "__version__",
    "extract_journal_reference",
    "extract_references_from_file",
    "extract_references_from_string",
    "extract_references_from_url",
)
